kahypar/partition/bin_packing/bin_packer.h

```diff
@@ -0,0 +1,124 @@
+class BinPacker final : public IBinPacker {
+ public:
+  BinPacker(Hypergraph& hypergraph, Context& context) : _hypergraph(hypergraph), _context(context) {}
+
+ private:
+  void prepackingImpl(const BalancingLevel level) override {
+    ASSERT(!_hypergraph.containsFixedVertices(), "No fixed vertices allowed before prepacking.");
+    ASSERT(level != BalancingLevel::STOP, "Invalid balancing level: STOP");
+
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    const HypernodeWeight max_bin_weight = floor(_context.initial_partitioning.current_max_bin_weight * (1.0 + _context.initial_partitioning.bin_epsilon));
+
+    if (level == BalancingLevel::heuristic) {
+      bin_packing::calculateHeuristicPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
+    } else if (level == BalancingLevel::guaranteed) {
+      for (size_t i = 0; i < static_cast<size_t>(_context.initial_partitioning.k); ++i) {
+        const HypernodeWeight lower = _context.initial_partitioning.perfect_balance_partition_weight[i];
+        const HypernodeWeight upper = _context.initial_partitioning.num_bins_per_part[i] * max_bin_weight;
+        HypernodeWeight& border = _context.initial_partitioning.upper_allowed_partition_weight[i];
+
+        // The performance of the prepacking algorithm depends on the perfect part weight, the allowed weight for each bin
+        // and the allowed part weight, which must be between those, i.e. lower <= border <= upper. If border has a value close
+        // to lower, the possibilities of the initial partitioning algorithm are rather restricted, but only few vertices need
+        // to be prepacked. On the other hand, if border has a value close to upper, almost every vertex will be prepacked,
+        // resulting in a bisection with a poor cut.
+        // To avoid this, we check for this case and adjust the allowed part weight if necessary.
+        if (upper - border < (border - lower) / 10) {
+          border = (lower + upper) / 2;
+          _context.partition.epsilon = static_cast<double>(border) / static_cast<double>(lower) - 1.0;
+        }
+      }
+
+      bin_packing::calculateExactPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
+    }
+  }
+
+  std::vector<PartitionID> twoLevelPackingImpl(const std::vector<HypernodeID>& nodes, const HypernodeWeight max_bin_weight) const override {
+    ASSERT(static_cast<size_t>(_context.partition.k) == _context.initial_partitioning.upper_allowed_partition_weight.size());
+
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    std::vector<PartitionID> parts(nodes.size(), -1);
+    TwoLevelPacker<BPAlgorithm> packer(rb_range_k, max_bin_weight);
+
+    if (_hypergraph.containsFixedVertices()) {
+      bin_packing::preassignFixedVertices<BPAlgorithm>(_hypergraph, nodes, parts, packer, _context.partition.k, rb_range_k);
```

Comment:
`bin_packing::` shouldn't be necessary here, since you are already in the `bin_packing` namespace.